pallets/ds-maps/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+        type Error = Error<T>;
+
+        // Events must be initialized if they are used by the pallet.
+        fn deposit_event() = default;
+
+        #[weight = <T as Trait>::WeightInfo::root_add()]
+        pub fn root_add(origin, 
+                        bounding_box: Box3D<Point3D<T::Coord>>,
+                        delta: T::Coord) -> dispatch::DispatchResult {
+            let who = ensure_signed(origin)?;
+            ensure!(<accounts::Module<T>>::account_is(&who, REGISTRAR_ROLE.into()), Error::<T>::NotAuthorized);
+            // Here more complex calculation for root dimensions and delta needed
+            let lat_dim = bounding_box.south_east.lat - bounding_box.north_west.lat;
+            ensure!(lat_dim.into() <= U9F23::from_num(1f64), Error::<T>::BadDimesions);
+            let lon_dim = bounding_box.south_east.lon- bounding_box.north_west.lon;
+            ensure!(lon_dim.into() <= U9F23::from_num(1f64), Error::<T>::BadDimesions);
+            ensure!(delta.into() <= U9F23::from_num(0.1f64), Error::<T>::InvalidData);
+
+            let id = TotalRoots::<T>::get();
+            let root = RootBoxOf::<T>::new(id, bounding_box, delta);
+            RootBoxes::<T>::insert(id, root);
+            TotalRoots::<T>::put(id + 1.into());
+            Self::deposit_event(RawEvent::RootCreated(id, who));
+            Ok(())
+        }
+        
+        /// Form index and store input to redzones, creates area struct if it doesnt exist
+        #[weight = <T as Trait>::WeightInfo::zone_add()]
+        pub fn zone_add(origin, 
+                        rect: Rect2D<Point2D<T::Coord>>,
+                        height: u16,
+                        root_id: T::RootId) -> dispatch::DispatchResult {
+            let who = ensure_signed(origin)?;
+            ensure!(<accounts::Module<T>>::account_is(&who, REGISTRAR_ROLE.into()), Error::<T>::NotAuthorized);
+            ensure!(RootBoxes::<T>::contains_key(root_id), Error::<T>::RootDoesNotExist);
+            ensure!(height > 1, Error::<T>::InvalidData);
+            // TODO calc required area in root from rect, rn no overlap checks
+            let area_id = Self::detect_touch(RootBoxes::<T>::get(root_id), rect.north_west);
+                        
+            let area = if AreaData::<T>::contains_key(root_id, area_id) {
+                AreaData::<T>::get(root_id, area_id)
+            } else {
+                AreaData::<T>::insert(root_id, area_id, Area::new(GREEN_AREA, 0));
+                Area::new(GREEN_AREA, 0)
+            };
+
+            ensure!(area.area_type == GREEN_AREA, Error::<T>::ForbiddenArea); 
+            let id = Self::form_index(root_id, area_id, area.child_amount); 
+            let zone = ZoneOf::<T>::new(id, rect, height);
+            RedZones::<T>::insert(id, zone);
+            AreaData::<T>::mutate(root_id, area_id, |ar| {
+                ar.child_amount += 1;
+            });
+            Self::deposit_event(RawEvent::ZoneCreated(root_id, area_id, id, who));
```

Comment:
What if the rect lies outside of root's bounding box (or intersects it)? Can't see any checks on that.